### PR TITLE
feature/login-session

### PR DESCRIPTION
**Pull request by GCerbelli**
## Summary

- Add login service
- Persist session
- Validate if user is previously logged

## Trello Card

[Nº 6.3 Funcionalidad de Login con Redux](https://trello.com/c/WxZIniaO/36-n-63-funcionalidad-de-login-con-redux)
[Nº 6.4 Persistencia de sesión](https://trello.com/c/ecCh2MMa/41-n-64-persistencia-de-sesi%C3%B3n)

## Screenshots
session-login

### CODE REVIEW

File: wbooks/src/services/BookService.js [hunk 1]
```diff
@@ -1,3 +1,3 @@
 import api from '../config/api';
 
-export const getBookList = () => api.get('/books');
+export const getBookList = (params) => api.get('/books', params);
```
Comment (marianozicavo) [inline wbooks/src/services/BookService.js:3]: I don't think you need this. Once you set the api headers it should be enough

Comment (GCerbelli) [reply wbooks/src/services/BookService.js:3]: I need it cause uid is not a header and it's required. If I set it at headers, I get a 401 status code response

Comment (GCerbelli) [reply wbooks/src/services/BookService.js]: Ok I move uid to headers and it works!

**Comment by GCerbelli** [comment]
At services, I have only type problems

At actions.js, I have only type problems too

And at store.js, I have "Cannot invoke an object which is possibly 'undefined'." at Reactotron.createEnhancer();
`if (__DEV__) {
  const reactotronMiddleware = Reactotron.createEnhancer();
  enhancers.push(reactotronMiddleware);
}`

File: wbooks/src/app/interfaces/api.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { HEADERS } from 'apisauce';
+
+import { User } from './authorization';
+
+export interface Headers {
```
Comment (marianozicavo) [inline wbooks/src/app/interfaces/api.ts:5]: Do you need this? It's the same as the one below

File: wbooks/src/redux/authorization/actions.ts [hunk 3]
```diff
@@ -7,12 +12,12 @@ export const actions = {
...
-  login: (credentials) => async (dispatch) => {
+  login: (credentials: Credentials) => async (dispatch: Dispatch) => {
     dispatch({ type: actions.LOGIN });
-    const response = await login(credentials);
+    const response: ApiResponse<LoginResponse, string> = await login(credentials);
     if (response.ok) {
-      const { client, 'access-token': token } = response.headers;
-      const user = response.data.data;
+      const { 'access-token': token, client } = response.headers as AuthorizationResponseHeaders;
+      const user = response.data?.data as User;
```
Comment (marianozicavo) [inline wbooks/src/redux/authorization/actions.ts:20]: Is this `as User` necessary? Why?

File: wbooks/src/services/AuthorizationService.ts [hunk 4]
```diff
@@ -1,16 +1,22 @@
 import AsyncStorage from '@react-native-community/async-storage';
+import { ApiResponse } from 'apisauce';
 
 import api from '../config/api';
 import STORAGE from '../constants/storage';
+import { User, Credentials } from '../app/interfaces/authorization';
+import { LoginResponse } from '../app/interfaces/api';
 
-export const login = (credentials) => api.post('/auth/sign_in', credentials);
+export const login = (credentials: Credentials): Promise<ApiResponse<LoginResponse, string>> =>
```
Comment (marianozicavo) [inline wbooks/src/services/AuthorizationService.ts:9]: You are defining `ApiResponse<LoginResponse, string>` here and in the actions.ts
I think you can define it only once

File: wbooks/tsconfig.json [hunk 5]
```diff
@@ -4,7 +4,7 @@
     /* Basic Options */
     "target": "esnext",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
     "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
-    "lib": ["es6"],                           /* Specify library files to be included in the compilation. */
+    "lib": ["es6", "es2019"],                 /* Specify library files to be included in the compilation. */
```
Comment (marianozicavo) [inline wbooks/tsconfig.json:7]: Try `"lib": ["esnext"],`

File: wbooks/src/redux/authorization/actions.ts (hunk 3 above)
Comment (GCerbelli) [reply wbooks/src/redux/authorization/actions.ts:20]: Cause it I don't have `as User`, I get this error at next line when I try to use `user`: 
`setAuthorizationData(user, token, client).then(() =>`
```
Argument of type 'User | undefined' is not assignable to parameter of type 'User'.
  Type 'undefined' is not assignable to type 'User'.
```

File: wbooks/tsconfig.json (hunk 5 above)
Comment (GCerbelli) [reply wbooks/tsconfig.json:7]: Good!

File: wbooks/src/redux/authorization/actions.ts (hunk 3 above)
Comment (marianozicavo) [reply wbooks/src/redux/authorization/actions.ts:19]: Let's try `const user = response.data!.data`

File: wbooks/src/redux/authorization/actions.ts [hunk 6]
```diff
@@ -14,7 +13,7 @@ export const actions = {
 export const actionCreators = {
   login: (credentials: Credentials) => async (dispatch: Dispatch) => {
     dispatch({ type: actions.LOGIN });
-    const response: ApiResponse<LoginResponse, string> = await login(credentials);
+    const response: LoginApiResponse = await login(credentials);
```
Comment (marianozicavo) [inline wbooks/src/redux/authorization/actions.ts:16]: You should be able to delete this part
```suggestion
    const response = await login(credentials);
```

File: wbooks/src/redux/authorization/actions.ts (hunk 3 above)
Comment (GCerbelli) [reply wbooks/src/redux/authorization/actions.ts:19]: Good, it ok now, thanks!

File: wbooks/src/redux/authorization/actions.ts (hunk 6 above)
Comment (GCerbelli) [reply wbooks/src/redux/authorization/actions.ts:16]: Yes, you're right

**Pull request merged**